server/game/gameSystems/DiscardCardsFromHand.ts

```diff
@@ -0,0 +1,84 @@
+    };
+
+    public override name = 'discard';
+    public override eventName = EventName.OnCardsDiscardedFromHand;
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    public override eventHandler(_event): void { }
+
+    public override getEffectMessage(context: TContext): [string, any[]] {
+        const properties = this.generatePropertiesFromContext(context);
+        return ['make {0} discard {1} cards', [properties.target, properties.amount]];
+    }
+
+    public override canAffect(playerOrPlayers: Player | Player[], context: TContext, additionalProperties = {}): boolean {
+        const players = Array.isArray(playerOrPlayers) ? playerOrPlayers : [playerOrPlayers];
+        return players.every((player) => {
+            const properties = this.generatePropertiesFromContext(context, additionalProperties);
+            const availableHand = player.hand.filter((card) => properties.cardCondition(card, context));
+
+            if (availableHand.length === 0 || properties.amount === 0) {
+                return false;
+            }
+            return super.canAffect(player, context);
+        });
+    }
+
+    public override queueGenerateEventGameSteps(events: any[], context: TContext, additionalProperties: Record<string, any> = {}): void {
+        const properties = this.generatePropertiesFromContext(context, additionalProperties);
+        (properties.target as Player[]).forEach((player) => {
+            const availableHand = player.hand.filter((card) => properties.cardCondition(card, context));
+            const amount = Math.min(availableHand.length, properties.amount);
+            if (amount <= 0) {
+                return;
+            }
+
+            if (amount >= availableHand.length) {
+                this.generateEventsForCards(availableHand, context, events);
+                return;
+            }
+
+            if (context.choosingPlayerOverride && context.choosingPlayerOverride !== player) {
+                this.generateEventsForCards(availableHand.slice(0, amount), context, events);
+                return;
+            }
+
+            context.game.promptForSelect(player, {
+                activePromptTitle: 'Choose ' + (amount === 1 ? 'a card' : amount + ' cards') + ' to discard',
+                context: context,
+                mode: TargetMode.Exactly,
+                numCards: amount,
+                locationFilter: Location.Hand,
+                controller: player === context.player ? RelativePlayer.Self : RelativePlayer.Opponent,
+                cardCondition: (card) => properties.cardCondition(card, context),
+                onSelect: (_player, cards) => {
+                    this.generateEventsForCards(cards, context, events);
+                    return true;
+                }
+            });
+        });
+    }
+
+    private generateEventsForCards(cards: Card[], context: TContext, events: any[]): void {
+        cards.forEach((card) => {
```

Comment:
Given this makes me wonder if we should revert back to having a single event that iterates through the cards and moves them. That would solve this issue, correct?

Reply:
Thinking on this more I wouldn't solve it cuz we still need to do as much as we can and that would be a successful even. So yeah back to brainstorming or doing one of the above

Reply:
We could make some extensions to where it would accomplish what we want - like right now there is a "resolved" state for GameEvent, we could add a "partially resolved" state as well without much issue.

The bigger problem with using one event is that it will cause inconsistencies for abilities that interact with discard, which I think I mentioned elsewhere. If the multi-discard case has `event.cards` as an array of the discarded cards but the card target system that discard a single card just uses `event.target`, but both emit the `onDiscardCards` event type, then any system that reads that event needs to know to check both properties.

There are ways around that as well, but imo that's the bigger issue here